web/src/search_pill.ts

```diff
@@ -89,45 +99,86 @@ function append_user_pill(
 
 const user_pill_operators = new Set(["dm", "dm-including", "sender"]);
 
+export function is_valid_pill(term: NarrowTerm): boolean {
+    switch (term.operator) {
+        case "has":
+            return ["image", "link", "attachment", "reaction"].includes(term.operand);
+        case "is":
+            return [
+                "dm",
+                "starred",
+                "mentioned",
+                "alerted",
+                "unread",
+                "resolved",
+                "followed",
+            ].includes(term.operand);
+        case "in":
+            return ["home", "all"].includes(term.operand);
+        case "id":
+            return message_store.get(Number.parseInt(term.operand, 10)) !== undefined;
+        case "channel":
+            return stream_data.get_sub(term.operand) !== undefined;
+        case "sender":
+        case "dm":
+        case "dm-including":
+            return term.operand
+                .split(",")
+                .every((email) => people.get_by_email(email) !== undefined);
+        default:
+            return true;
+    }
+}
```

Comment:
This seems to not be exhaustive; the `streams:` operator should limit what's valid here.

Reply:
Oh oops, I thought `streams` couldn't reach here and only `channel`, but I found a way where it can, thanks!

Reply:
> Might be worth putting this in as Filter.is_valid_search_pill, defined just after Filter.parse ; you could do that as a prep PR that defines the function, with tests, while unused, just to get logic integrated separately from the main PR.

I've set up that as a prep commit and made a PR for it here #30678